Source/WTF/wtf/text/StringConcatenate.h

```diff
@@ -148,104 +148,38 @@ template<> class StringTypeAdapter<char*, void> : public StringTypeAdapter<const
-    template<typename CharacterType> void writeTo(CharacterType* destination) const { StringImpl::copyCharacters(destination, characters(), length()); }
-
-private:
-    const CharTypeForString* characters() const
-    {
-        return reinterpret_cast<const CharTypeForString*>(m_vector.data());
-    }
-
-    const Vector<CharType, N>& m_vector;
-};
-
-template<> class StringTypeAdapter<StringImpl*, void> {
-public:
-    StringTypeAdapter(StringImpl* string)
-        : m_string { string }
-    {
-    }
-
-    unsigned length() const { return m_string ? m_string->length() : 0; }
-    bool is8Bit() const { return !m_string || m_string->is8Bit(); }
-    template<typename CharacterType> void writeTo(CharacterType* destination) const
-    {
-        StringView { m_string }.getCharacters(destination);
-        WTF_STRINGTYPEADAPTER_COPIED_WTF_STRING();
-    }
-
-private:
-    StringImpl* const m_string;
-};
-
-template<> class StringTypeAdapter<AtomStringImpl*, void> : public StringTypeAdapter<StringImpl*, void> {
-public:
-    StringTypeAdapter(AtomStringImpl* string)
-        : StringTypeAdapter<StringImpl*, void> { static_cast<StringImpl*>(string) }
-    {
-    }
-};
-
-template<> class StringTypeAdapter<String, void> : public StringTypeAdapter<StringImpl*, void> {
-public:
-    StringTypeAdapter(const String& string)
-        : StringTypeAdapter<StringImpl*, void> { string.impl() }
-    {
-    }
-};
-
-template<> class StringTypeAdapter<AtomString, void> : public StringTypeAdapter<String, void> {
+template<typename CharacterType, size_t N>
+class StringTypeAdapter<CharacterType[N], void> : public StringTypeAdapter<const CharacterType*, void> {
 public:
-    StringTypeAdapter(const AtomString& string)
-        : StringTypeAdapter<String, void> { string.string() }
+    StringTypeAdapter(const CharacterType(&characters)[N])
+        : StringTypeAdapter<const CharacterType*, void> { characters }
     {
     }
 };
 
-template<> class StringTypeAdapter<StringImpl&, void> {
+template<> class StringTypeAdapter<StringView, void> {
 public:
-    StringTypeAdapter(StringImpl& string)
+    StringTypeAdapter(const StringView& string)
         : m_string { string }
     {
     }
 
     unsigned length() const { return m_string.length(); }
     bool is8Bit() const { return m_string.is8Bit(); }
-    template<typename CharacterType> void writeTo(CharacterType* destination) const
-    {
-        StringView { m_string }.getCharacters(destination);
-        WTF_STRINGTYPEADAPTER_COPIED_WTF_STRING();
-    }
+    template<typename CharacterType> void writeTo(CharacterType* destination) const { m_string.getCharacters(destination); }
 
 private:
-    StringImpl& m_string;
+    StringView m_string;
 };
 
-template<> class StringTypeAdapter<AtomStringImpl&, void> : public StringTypeAdapter<StringImpl&, void> {
+// Array types in the following partial specialization are disallowed in order to avoid implicit constructors
```

Comment:
Which String implicit ctor is used?
Does String have an implicit ctor for a character array?

Reply:
UserAgentGLib.cpp is compiled only by GTK and WPE. Let's add a test case in https://github.com/WebKit/WebKit/blob/main/Tools/TestWebKitAPI/Tests/WTF/StringConcatenate.cpp .

Reply:
> Which String implicit ctor is used?
> Does String have an implicit ctor for a character array?

Via `StaticStringImpl`.

Reply:
We have `""_s` operator. I think we should disallow constructing a String from a character array. Right?

Reply:
I looked into this. Interesting problem. But, I confirmed that StringView can't be constructed a string literal. I think the usage of is_constructible_v can be impoved. How about thi patch?
https://bugs.webkit.org/show_bug.cgi?id=250017#c10

Reply:
> We have `""_s` operator. I think we should disallow constructing a String from a character array. Right?

I would say yes, even if it's a pretty big restriction. I imagine it happens a lot in the code.

Reply:
> I looked into this. Interesting problem. But, I confirmed that StringView can't be constructed from a string literal. I think the usage of is_constructible_v can be impoved. How about thi patch?
> https://bugs.webkit.org/show_bug.cgi?id=250017#c10

Thanks for the proposal, I'll investigate it.

I have a few other changes queued up behind this one, including again supporting string literals in StringView, but this time with enforced length computation (i.e. no more dependence on the array size value). I'll notify you on relevant PRs as I open them.

Reply:
> including again supporting string literals in StringView

Is it a good idea? Why do we have ASCIILiteral?  StringView explicitly disallows it.
https://github.com/WebKit/WebKit/blob/1b38f29b8d045a32314964794baa0094954e3b69/Source/WTF/wtf/text/StringView.h#L198-L200

Reply:
It sounds nice, but it's against https://github.com/WebKit/WebKit/pull/1040 .
Let's make a consensus rather than edit warring.

Reply:
OK. I agree supporting raw string literals (i.e. fixed-length LChar or UChar arrays) isn't the best option, even if it's trivial to support. I can add them in the future if there's actual benefits with them being present.

Reply:
I asked them. This is the reason why we should disallow `const char*` for StringView and makeString.
https://bugs.webkit.org/show_bug.cgi?id=240942#c11